common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -0,0 +1,883 @@
+import org.apache.spark.network.protocol.Encoders;
+import org.apache.spark.network.shuffle.protocol.ExecutorShuffleInfo;
+import org.apache.spark.network.shuffle.protocol.FinalizeShuffleMerge;
+import org.apache.spark.network.shuffle.protocol.MergeStatuses;
+import org.apache.spark.network.shuffle.protocol.PushBlockStream;
+import org.apache.spark.network.util.JavaUtils;
+import org.apache.spark.network.util.NettyUtils;
+import org.apache.spark.network.util.TransportConf;
+
+/**
+ * An implementation of {@link MergedShuffleFileManager} that provides the most essential shuffle
+ * service processing logic to support push based shuffle.
+ */
+public class RemoteBlockPushResolver implements MergedShuffleFileManager {
+
+  private static final Logger logger = LoggerFactory.getLogger(RemoteBlockPushResolver.class);
+  @VisibleForTesting
+  static final String MERGE_MANAGER_DIR = "merge_manager";
+
+  private final ConcurrentMap<String, AppPathsInfo> appsPathInfo;
+  private final ConcurrentMap<AppShuffleId, Map<Integer, AppShufflePartitionInfo>> partitions;
+
+  private final Executor directoryCleaner;
+  private final TransportConf conf;
+  private final int minChunkSize;
+  private final ErrorHandler.BlockPushErrorHandler errorHandler;
+
+  @SuppressWarnings("UnstableApiUsage")
+  private final LoadingCache<File, ShuffleIndexInformation> indexCache;
+
+  @SuppressWarnings("UnstableApiUsage")
+  public RemoteBlockPushResolver(TransportConf conf) {
+    this.conf = conf;
+    this.partitions = Maps.newConcurrentMap();
+    this.appsPathInfo = Maps.newConcurrentMap();
+    this.directoryCleaner = Executors.newSingleThreadExecutor(
+      // Add `spark` prefix because it will run in NM in Yarn mode.
+      NettyUtils.createThreadFactory("spark-shuffle-merged-shuffle-directory-cleaner"));
+    this.minChunkSize = conf.minChunkSizeInMergedShuffleFile();
+    CacheLoader<File, ShuffleIndexInformation> indexCacheLoader =
+      new CacheLoader<File, ShuffleIndexInformation>() {
+        public ShuffleIndexInformation load(File file) throws IOException {
+          return new ShuffleIndexInformation(file);
+        }
+      };
+    indexCache = CacheBuilder.newBuilder()
+      .maximumWeight(conf.mergedIndexCacheSize())
+      .weigher((Weigher<File, ShuffleIndexInformation>) (file, indexInfo) -> indexInfo.getSize())
+      .build(indexCacheLoader);
+    this.errorHandler = new ErrorHandler.BlockPushErrorHandler();
+  }
+
+  /**
+   * Given the appShuffleId and reduceId that uniquely identifies a given shuffle partition of an
+   * application, retrieves the associated metadata. If not present and the corresponding merged
+   * shuffle does not exist, initializes the metadata.
+   */
+  private AppShufflePartitionInfo getOrCreateAppShufflePartitionInfo(
+      AppShuffleId appShuffleId,
+      int reduceId) {
+    Map<Integer, AppShufflePartitionInfo> shufflePartitions =
+      partitions.computeIfAbsent(appShuffleId, id -> Maps.newConcurrentMap());
```

Comment:
Another effect of breaking the map into a 2-level map.
After we delete the first level map from `partitions` during merge finalization, if there is another block received for the same shuffle, we would then repopulate the first level map with an empty map, even though we have already finalized this shuffle.
It's possible that we would keep these empty maps in the partitions map until the application finishes.

I feel that breaking the previous 1-level map into a 2-level map has quite some side effects.
If what we want is to speedup the iteration during shuffle finalization, would keep a separate map for (appId, shuffleId) -> reduceId also address the need?
This way, we do not break the `partitions` map into a 2-level map which lead to these side effects.

Reply:
One more aspect to consider.
In order to support YARN NM restart, we need to store the information tracked inside the `partition` map into a levelDB state store, similarly to what `ExternalShuffleBlockResolver` is doing.
Having 1-level map would make storing the information inside levelDB much easier compared with a 2-level map.

Reply:
CC @Ngone51 @mridulm

Reply:
That is a good point @Victsm, the only concern with maintaining a second data structure is potential for divergence (or need to wrap use of either in critical section).

Thoughts @Ngone51 ?

Reply:
Thinking more, this specific issue can be resolved by introducing the new map only if it will result in addition to it.
I have sketched a probable flow to @otterc, but in a nutshell, it will be :

```
AtomicReference<AppShufflePartitionInfo> result = new AtomicReference<>(null);
Map<Integer, AppShufflePartitionInfo> shufflePartitions =
      partitions.computeIfAbsent(appShuffleId, id -> {
        // rename
        AppShufflePartitionInfo info = createAppShufflePartitionInfo(<params>);
        if (null != info) {
          Map<Integer, AppShufflePartitionInfo> map = Maps.newConcurrentMap();
          map.put(reduceId, info);
          result.set(info);
          return map;
        } else {
          return null;
        }
      });

if (null != result.get()) return result.get();
return null == shufflePartitions ? null : shufflePartitions.computeIfAbsent(reduceId, key -> createAppShufflePartitionInfo(<params>);
}     
```

Reply:
Another way I'm thinking of is, maybe, we can do like:

```scala
private AppShufflePartitionInfo getOrCreateAppShufflePartitionInfo(
      AppShuffleId appShuffleId,
      int reduceId) {
  if (!partitions.containsKey(appShuffleId) && getMergedShuffleDataFile(appShuffleId, reduceId).exists) {
     // finalized
    return null;
  } else {
    // get existing or reate new AppShufflePartitionInfo
    ...
  }
}

```

Reply:
> Another way I'm thinking of is, maybe, we can do like:
> 
> ```scala
> private AppShufflePartitionInfo getOrCreateAppShufflePartitionInfo(
>       AppShuffleId appShuffleId,
>       int reduceId) {
>   if (!partitions.containsKey(appShuffleId) && getMergedShuffleDataFile(appShuffleId, reduceId).exists) {
>      // finalized
>     return null;
>   } else {
>     // get existing or create new AppShufflePartitionInfo
>     ...
>   }
> }
> ```

This could work if the entry for appShuffleId from partitions is removed after all the files of all partitions are closed in finalizeShuffleMerge.

Reply:
I have made this change